### PR TITLE
[MFMA] Support 64x4 and 4x64 tile size

### PR DESCRIPTION
**Pull request by binarman**
This PR enables two new MxN tile sizes: 64 x 4 and 4 x 64. Both of them uses mfma 4x4 instructions.

### CODE REVIEW

File: test/TritonGPU/accelerate-matmul-cdna1.mlir [hunk 1]
```diff
@@ -0,0 +1,800 @@
+// RUN: (! triton-opt %s -split-input-file --tritonamdgpu-accelerate-matmul=arch-generation-name=gfx908 --mlir-pass-pipeline-crash-reproducer=%t 2>/dev/null) | FileCheck --check-prefixes=CHECK %s
```
Comment (alefimov-amd) [inline test/TritonGPU/accelerate-matmul-cdna1.mlir:1]: This file is generated,
I will add scripts in next PR, just don't want to add too much things in one PR

File: lib/Dialect/TritonGPU/Transforms/Utility.cpp [hunk 2]
```diff
@@ -671,131 +671,180 @@ auto getMfmaInsnGroupAttrMap = []() -> const MfmaInsnGroupMap & {
...
-      {{16, MfmaTypeId::Fp32TyId, 2},
+      {{16, 16, MfmaTypeId::Fp32TyId, 2},
        {16, 16, 4, 1, ROCDL::mfma_f32_16x16x4f32::getOperationName()}},
-      {{16, MfmaTypeId::Fp32TyId, 3},
+      {{16, 16, MfmaTypeId::Fp32TyId, 3},
        {16, 16, 4, 1, ROCDL::mfma_f32_16x16x4f32::getOperationName()}},
       // mfma_f32_4x4x1f32
-      {{4, MfmaTypeId::Fp32TyId, 1},
+      {{4, 4, MfmaTypeId::Fp32TyId, 1},
        {4, 4, 16, 1, ROCDL::mfma_f32_4x4x1f32::getOperationName()}},
```
Comment (zhanglx13) [inline lib/Dialect/TritonGPU/Transforms/Utility.cpp:689]: I think I made a mistake here. 
It should be 4 , 4, 1, 1 instead of 4, 4, 16, 1

Comment (zhanglx13) [reply lib/Dialect/TritonGPU/Transforms/Utility.cpp:689]: Oh, I realized this is for (4x16) x (16x4) --> 4x4. nvm

File: lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandMFMA.cpp [hunk 3]
```diff
@@ -132,6 +132,7 @@ swizzleIndexes(ConversionPatternRewriter &rewriter, Location loc, Value row,
  * @param smemStrides strides in LDS tensor
  * @param loadVecSize number of elements loaded by one operation
  * @param iNonKDim non-K dimension of dot operand
+ * @param iKDim non-K dimension of dot operand
```
Comment (zhanglx13) [inline lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandMFMA.cpp:135]: should be K dimension

File: lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVMBase.h [hunk 4]
```diff
@@ -1233,10 +1239,15 @@ class ConvertTritonGPUOpToLLVMPatternBase {
     auto warpsPerCTA = mfmaLayout.getWarpsPerCTA();
 
     SmallVector<unsigned> numWarpsPerDim(2);
+    unsigned mDim = mfmaLayout.getMDim();
+    unsigned nDim = mfmaLayout.getNDim();
+    assert((mDim == nDim && (mDim == 32 || mDim == 16 || mDim == 4)) ||
```
Comment (zhanglx13) [inline lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVMBase.h:1244]: Do we really need this kind of assert everywhere? Can we only check once at the earliest from the codegen?

**Review by zhanglx13 — APPROVED** [review]
I think this PR is good to go.

**Comment by zhanglx13** [comment]
@scxiao @vgokhale 
The heuristic for picking mfma instruction size is as follows
1. If the result tile shape is larger than 32x32, pick mfma32
2. If the tile shape is smaller than 32x32 but larger than 16x16, pick mfma16
3. if the tile shape is smaller than 4x64 or 64x4, pick mfma4x4
4. Otherwise, pick mfma4x64 or mfma64x4

However, in the case of FA decode kernel, the tile shape is 16x128. And mfma16 will be picked according to the heuristic.
The tile shape refers to the result tensor shape of `tt.dot`. This heuristic does not take num_warps into consideration. But we do not have warp layout information when choosing mfma dimensions. Therefore, the only solution here is enable some user input to enforce the choice of mfma4x64 here.

@alefimov-amd In the next PR, can you change `chooseMfmaDimensions` to pick 4x64 or 64x4 based on the tile shape when `matrix_instr_nonkdim` is 4?

**Comment by scxiao** [comment]
> chooseMfmaDimensions

Can you specify a unique value of `chooseMfmaDimensions` to choose mfma4x64 and mfma64x4, like `464` and `644`?

**Review by scxiao — APPROVED** [review]

File: lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVMBase.h (hunk 4 above)
Comment (binarman) [reply lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVMBase.h:1244]: I agree, this code is redundant most of the time. It just helps add new layouts and do not forget anything crucial.

You can set new (unsupported) m/n combination in accelerate matmul pass, run tests and see where these asserts fire.

**Comment by binarman** [comment]
> In the next PR, can you change chooseMfmaDimensions to pick 4x64 or 64x4 based on the tile shape when matrix_instr_nonkdim is 4?

sure

> Can you specify a unique value of chooseMfmaDimensions to choose mfma4x64 and mfma64x4, like 464 and 644?

This is useful idea, thank you!

File: lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandMFMA.cpp (hunk 3 above)
Comment (binarman) [reply lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandMFMA.cpp:135]: I want to make another PR with some fixes, let me include this fix there.

**Pull request merged**